slurm_ops_manager/infiniband.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python3
+"""Infiniband Class."""
+
+import logging
+import os
+from pathlib import Path
+import shlex
+import shutil
+import subprocess
+
+from ops.framework import Object, StoredState
+
+from slurm_ops_manager.utils import operating_system
+
+
+logger = logging.getLogger()
+
+BASE_DIR = os.path.dirname(os.path.abspath(__file__))
+
+class Infiniband(Object):
+    """Responsible for Infiniband operations."""
+
+    _stored = StoredState()
+
+    def __init__(self, parent, key):
+        """Initialize class."""
+        super().__init__(parent, key)
+
+        self._stored.set_default(ib_installed=False,
+                                 ib_repo_configured=False,
+                                 ib_custom_repo=False,
+                                 ib_repo_path=str(), # it doesn't accept Path
+                                 ib_default_repo=str(),
+                                 ib_package_name='mlnx-ofed-all')
+
+        self._ib_systemd_service = 'openibd.service'
+        self._template_dir = Path(BASE_DIR) / 'templates'
+        self._operating_system = operating_system()
+
+        # setup repositories based on official Mellanox docs
+        if self._operating_system == 'ubuntu':
+            self._stored.ib_repo_path = "/etc/apt/sources.list.d/infiniband.list"
+            self._stored.ib_default_repo = str(self._template_dir) + "/mellanox_repo_ubuntu.list"
+        elif self._operating_system == 'centos':
+            self._stored.ib_repo_path = "/etc/yum.repos.d/infiniband.repo"
+            self._stored.ib_default_repo = str(self._template_dir) + "/mellanox_repo_centos7.repo"
+        else:
+            logger.error(f'#### Unsupported OS to install infiniband: {self._operating_system}')
+
+    @property
+    def repository(self):
+        """Return the repository used for Infiniband drivers."""
+        if self._stored.ib_repo_configured:
+            return Path(self._stored.ib_repo_path).read_text()
+        else:
+            return "Repository not set up"
+
+    @repository.setter
+    def repository(self, repo: str):
+        """Set a custom repository to install Infiniband drivers.
+
+        If the new repository string is empty, re-sets to the default.
+        """
+        if repo:
+            logger.debug(f'#### Infiniband - setting custom repo: {repo}')
+            Path(self._stored.ib_repo_path).write_text(repo)
+        else:
+            logger.debug(f'#### Infiniband - setting default repo')
+            shutil.copyfile(self._stored.ib_default_repo,
+                            self._stored.ib_repo_path)
+
+            if self._operating_system == 'ubuntu':
+                logger.debug('#### Infiniband - adding GPG keys')
+                key_url = 'http://www.mellanox.com/downloads/ofed/RPM-GPG-KEY-Mellanox'
```

Comment:
I know this is only used here ....

How do you feel about defining all strings as class level variables, instead of inline where they are used in the code?

Reply:
I actually don't know how I feel about what I just said either .... I'm not too opinionated here. Feel free to leave it as is, or pull this stuff up to the top. Either way works for me.

Reply:
I prefer to have variables with smaller scope possible. This makes it easier to see where stuff is used.